File: Moving_chunks/Moving_chunks.py
spam = input().split(' ')
n = int(spam[0]) # количество чанков на кластере
m = int(spam[1]) # количество серверов
q = int(spam[2]) # количество запросов

spam = input().split(' ')
chank = [] # список с серверами, на которых расположены чанки
for i in spam:
    chank.append(int(i))

inquiries = [] # список для запросов
for i in range(q):
    inquiries.append(input())

for inquiry in inquiries:
    spam = inquiry.split(' ')
    server_from = int(spam[0])
    server_to = int(spam[1])
    chank_b = int(spam[2]) - 1
    chank_e = int(spam[3])
    s = set(chank[chank_b : chank_e]) # множество из серверов, на которых расположены чанки, которые нужно переместить в запросе
    if len(s) == 1 and server_from in s:
        # Если перемещаемые в запросе чанки находятся на нужном сервере,
        # чанки перемещаются
        for i in range(chank_b, chank_e):
            chank[i] = server_to
        print(1)
    else:
        # Если не все перемещаемые чанки находятся на нужном сервере,
        # чанки не перемещаются
        print(0)

File: Moving_chunks/Moving_chunks.cpp
#include <iostream>
#include <set>

int main(int argc, char* argv[])
{
	int n; // количество чанков на кластере
	int m; // количество серверов
	int q; // количество запросов
	std::cin >> n >> m >> q;
	int* chank = new int[n]; // массив с серверами, на которых расположены чанки
	int i, j;
	// Считывается массив с серверами, на которых расположены чанки
	for (i = 0; i < n; i++)
	{
		std::cin >> chank[i];
	}
	int** inquiries; // массив с запросами
	inquiries = new int* [q];
	// Считывается массив с запросами
	for (i = 0; i < q; i++)
	{
		inquiries[i] = new int[4];
		for (j = 0; j < 4; j++)
		{
			std::cin >> inquiries[i][j];
		}
		inquiries[i][2]--;
		inquiries[i][3]--;
	}
	// Выполняются запросы
	for (i = 0; i < q; i++)
	{
		int server_from = inquiries[i][0]; // с какого сервера нужно переместить чанки
		int	server_to = inquiries[i][1]; // на какой сервер нужно переместить чанки
		int	chank_b = inquiries[i][2]; // какой номер первого чанка, который нужно переместить
		int	chank_e = inquiries[i][3]; // какой номер последнего чанка, который нужно переместить
		for (j = chank_b; j <= chank_e; j++)
		{
			if (chank[j] != server_from)
			{
				// Если не все перемещаемые чанки находятся на нужном сервере, чанки не перемещаются
				std::cout << 0 << std::endl;
				break;
			}
		}
		if (j > chank_e)
		{
			// Если перемещаемые в запросе чанки находятся на нужном сервере, чанки перемещаются
			for (j = chank_b; j <= chank_e; j++)
			{
				chank[j] = server_to;
			}
			std::cout << 1 << std::endl;
		}
	}
	for (i = 0; i < q; i++)
	{
		delete(inquiries[i]);
	}
	delete(inquiries);
}